answers.md

```diff
@@ -0,0 +1,116 @@
+### Chapter 12 Self Test
+
+  1) _Enumeration constants are said to be **self-typed**. What does
+      this mean?_
+
+Enumeration constants are typed by their enclosing class. For example,
+in the enum:
+
+```java
+enum Creature {
+	PLAYER, NPC, MONSTER
+}
+```
+
+`MONSTER` is of the type `Creature`.
+
+  2) _What class do all enumerations automatically inherit?_
+
+`java.lang.Enum`.
+
+  3) _Given the following enumeration, write a program that uses `values()`
+      to show a list of the constants and their ordinal values._
+
+```java
+enum Tools {
+	SCREWDRIVER, WRENCH, HAMMER, PLIERS
+}
+```
+
+```java
+Arrays.stream(Tools.values()).forEach(
+	t -> System.out.printf("%s: %d%n", t.name(), t.ordinal()) );
+```
+
+  4) _The traffic light simulation developed in **Try This 12-1** can be
+      improved with a few simple changes that take advantage of an enumeration's
+      class features. In the version shown, the duration of each color was
+      controlled by the `TrafficLightSimulator` class by hard-coding these
+      values into the `run()` method. Change this so that the duration of each
+      color is stored by the constants in the `TrafficLightColor` enumeration.
+      To do this, you will need to add a constructor, a private instance
+      variable, and a method called `getDelay()`. After making these changes,
+      what improvements do you see? On your own, can you think of other
+      improvements? (Hint: Try using ordinal values to switch light colors
+      rather than relying on a `switch` statement.)_
+
+See [TrafficLightDemo](src/org/java/training/TrafficLightDemo.java).
+
+  5) _Define boxing and unboxing. How does autoboxing/unboxing affect these
+      actions?_
+
+Boxing is a mechanism for wrapping Java primitives in Objects to provide OO
+features and allow primitives to be used in Java's limited attempt at generic
+programming. Unboxing is the process of getting a primitive value from a boxed
+Object reference. Autoboxing/unboxing affects these by removing the boilerplate
+associated with these conversions, making them usable.
```

Comment:
I'm being _pedantic_ again, but you would have gotten extra credit if you had explicitly said that Autoboxing/unboxing did the conversion automatically (thus removing the associated boilerplate).